### ISSUE TITLE
Memory access violation on "A connection error occurred" exception

### ISSUE THREAD

**Issue by FunnyPixel** [D7EAD/liboai#16]
### Describe the bug

This simple example leads to a crash with some memory access violation in xmemory.h ( I am on windows)

```
#include "liboai.h"

using namespace liboai;

int main() {
  OpenAI oai;
  if (oai.auth.SetKeyEnv("OPENAI_API_KEY")) {
    try {
      Response response = oai.Image->create(
        "a siamese cat!"
      );
    }
    catch (std::exception& e) {
      std::cout << e.what() << std::endl;
    }
    
    ...
  }
}
```

The lib crashes  somewhere after 

  ```
 throw liboai::exception::OpenAIException(
			"A connection error occurred",
			liboai::exception::EType::E_CONNECTIONERROR,
			"liboai::Response::CheckResponse()"
		);
```
is called. I can't see the exact error location on the library side in call stack, but eventually it leads to a crash in xmemory.h.

MSVS spits this:

>   Exception thrown at 0x00007FF8D13ECD29 in TestApp.exe: Microsoft C++ exception: liboai::exception::OpenAIException at memory location 0x00000046F09DBFC0.
> Exception thrown at 0x00007FF7DD1C82C5 in TestApp.exe: 0xC0000005: Access violation reading location 0xFFFFFFFFFFFFFFFF.

In my case I have no credits left in OpenAI API. Have you ever tested that case?

 

### To Reproduce

Run the code I provided below. But maybe it is also a special case related to the fact the account has no credits left.
 

### Code snippets

```C++
int main()
{
    OpenAI oai;
    if (oai.auth.SetKey("my key goes here"))
    {
        try {

           /* Response res = oai.Image->create(
                "A snake in the grass!",
                1,
                "256x256"
            );*/

            Response response = oai.Image->create(
                "a siamese cat!"
            );

            std::cout << response["data"][0]["url"] << std::endl;
        }
        catch (std::exception& e)
        {
            std::cout << e.what() << std::endl;
            std::quick_exit(EXIT_FAILURE);
        }
   }

    std::cout << "Hello World!\n";
    return 0;
}
```

### OS

Windows

### Library version

3.0.0

**Comment by D7EAD** [comment]
This could possibly be a special case with the account having no credits which I have, unfortunately, not tested. 

Furthermore, the code `E_CONNECTIONERROR` is thrown when a returned response code is zero, which usually means it is unset by cURL.

So, this could either be a special case for a lack of tokens, an issue with cURL, or a mixture of both. However, I have tested all snippets when I implemented the new cURL wrapper, so I lean towards it being a special case. 

I'll surely take a look at this soon and attempt to fix it. In the meantime, feel free to look into this yourself.

**Comment by D7EAD** [comment]
On second glance, the case of exceeding your quota shouldn't be an issue. It would return a 429 response code and be caught as a rate limit exception, assuming the request was successful. 

I'll keep you updated.

**Comment by sasmaster** [comment]
Hi. The above user is me from another account. So, I have charged the account with money and I see nothing changes. Same crash.
I have tried also openai-cpp now and the error I am getting there is related to curl:

>  OpenAI curl_easy_perform() failed: Unsupported protocol

Can it be somehow related to curl? Any specific version I should use or a feature to enable in the build? I build libcurl from the source,so may be some protocols are off by default.
Thanks.

**Comment by D7EAD** [comment]
Thanks for looking more into this and narrowing down possible sources. I will have time to investigate, and hopefully fix, this tomorrow (4/6).

**Comment by D7EAD** [comment]
> Also, I found that specifying organization name (verified it is correct) returns :
> 
> > "error": {
> > "code": "invalid_organization",
> > "message": "No such organization: Myorg.",
> > "param": null,
> > "type": "invalid_request_error"
> > }
> 
> "Myorg" is not supposed to be with "." at the end. It gets appended somewhere.

I've investigated this supposed behavior regarding organization IDs and have confirmed that there is no period appended anywhere. The period is appended to the name by the OpenAI API response when an incorrect organization ID is provided.

Organization IDs are in the format: `org-XXXXXXXXXXXXXXXXXXXXXXXX`. So, for instance:

```cpp
oai.auth.SetOrganization("org-L336fOLglLPXi4XA2D7wxvD5")
```

Regarding the other reported issues, I will update once I've investigated.

**Comment by sasmaster** [comment]
Oh, I thought organization should be the name which I have registered  under the account. Haven't used that ID you're talking about. Then we got that crush at the end of the response.

**Comment by D7EAD** [comment]
> Oh, I thought organization should be the name which I have registered under the account. Haven't used that ID you're talking about. Then we got that crush at the end of the response.

Yes, I will post an update regarding this today. I have had some trouble replicating the behavior of the crash you have described, so I will be adding additional safeguards and checks today to hopefully handle any errors that may lead to said crash.

**Comment by D7EAD** [comment]
Hi.

I've created another branch (#17) to which I am pushing, and will push further, updates regarding this issue to. Feel free to test it and let me know if the issues regarding crashes have been resolved or, at least, provide clearer hints as to what is causing them. 

Thanks.

**Comment by sasmaster** [comment]
Going to check it.
Thanks.

**Comment by sasmaster** [comment]
So I compiled and ran a test with that branch you have created. I see you added SSL activation check which I pass I made sure to build curl with OpenSSL enabled. Then it failed at the memory access error as before. Here is something I have been able to find:
In the call stack (see image):

![image](https://user-images.githubusercontent.com/438400/230736263-af863e32-85c6-456b-9649-68900be53031.png)

Yellow marked is a string which contains  `"https://api.openai.com/v1/chat/completions"`
Next call is `std::string:: _Tidy_deallocate`  is where the crash starts which leads to wreck in xmemory code.
Beyond that I cannot tell why it happens. It is weird as usually std::string internal memory management is self contained. Maybe string views you're using cause that?

**Comment by D7EAD** [comment]
> So I compiled and ran a test with that branch you have created. I see you added SSL activation check which I pass I made sure to build curl with OpenSSL enabled. Then it failed at the memory access error as before. Here is something I have been able to find: In the call stack (see image):
> 
> ![image](https://user-images.githubusercontent.com/438400/230736263-af863e32-85c6-456b-9649-68900be53031.png)
> 
> Yellow marked is a string which contains `"https://api.openai.com/v1/chat/completions"` Next call is `std::string:: _Tidy_deallocate` is where the crash starts which leads to wreck in xmemory code. Beyond that I cannot tell why it happens. It is weird as usually std::string internal memory management is self contained. Maybe string views you're using cause that?
> 
> Before the crash I am able to see the response from GPT ,so network communication takes place fine.
> 
> Btw, are you using windows? If you do maybe you can send me your libcurl lib or I can upload mine to exclude the possibility the problem is with curl (maybe some compiler flags cause binary incompatibility).

Hmm, this is certainly helpful. 

In regard to `string_view`, if I remember correctly, all instances of it are not stored and are either:
1. binded to compile-time character arrays, or
2. sent to the constructor of `std::string`

Do you happen to be compiling in Debug mode instead of Release mode?

**Comment by sasmaster** [comment]
Some more details:

It fails on return of this:

 
```
return liboai::Response {
		std::move(this->url_str),
		std::move(this->content),
		std::move(this->status_line),
		std::move(this->reason),
		this->status_code,
		this->elapsed
	};
```

It crushes on the string memory dealloc after the return of the new Response object and before this returns:

```
liboai::Response liboai::netimpl::Session::Complete() {
	this->hasBody = false;
	return this->BuildResponseObject();
}
```
Very weird. I tried to create another liboai::Response constructor which accepts not rvalue  strings but const references but same thing.

**Comment by sasmaster** [comment]
I need it in Debug mode as I am integrating into another C++ app. But this test is a simple standalone 'hello world' exe.

**Comment by D7EAD** [comment]
> I need it in Debug mode as I am integrating into another C++ app. But this test is a simple standalone 'hello world' exe.

It's very likely that it being in Debug mode is causing the xmemory crash. This has happened to me countless times when debugging apps and is a lasting issue with the MS standard library implementation, according to how many times it has been reported on the VS forum.

**Comment by D7EAD** [comment]
Okay, I can confirm I have isolated this issue to Debug mode builds.

**Comment by sasmaster** [comment]
Appreciate. I am also trying to learn why it crushes. It is an interesting CPP anomaly.

**Comment by D7EAD** [comment]
> Appreciate. I am also trying to learn why it crushes. It is an interesting CPP anomaly.

Is it possible you are linking both Release and Debug builds of dependencies together? If you are build in Debug mode, then all dependencies should have been built in Debug mode, such as cURL, the runtime, and other dependencies.

**Comment by sasmaster** [comment]
Linking all in Debug. OpenSSL I am using only .dll since libcurl is also dynamic library.

**Comment by D7EAD** [comment]
> Linking all in Debug. OpenSSL I am using only .dll since libcurl is also dynamic library.

Alright, I'll be working on this and will post updates here and push them to the 3.0.1 branch. Feel free to attempt to fix it yourself as well and, if you do, feel free to open a PR for the 3.0.1 branch.

**Comment by sasmaster** [comment]
Nailed it. This  is the code that corrupts the internal string buffer:

`e[2] = curl_easy_getinfo(this->curl_, CURLINFO_EFFECTIVE_URL, &this->url_str);`

Reading curl doc for this function it all makes sense. The function will fill the buffer with data. Now, you can't do that with std::string as manipulations of its internal memory outside the container may lead to all kind of troubles. Enough that the filled data is bigger than the buffer or no null char is added by the filling logic - it all explodes. I also reviewed your whole API and here just a couple of friendly tips: I am more orthodox in C++ dev so of course I cannot urge you to write less meta-programming,but still,I would reduce amount of modern C++ wherever possible, as it would make the code more readable and easier to maintain. Template classes with variadic arg forwarding are extremely hard to debug. If there are no endless variations for function signatures I would just implement those several overloaded methods instead. Your lib is otherwise robust but as I said, less templatization would make it easier to use and more portable.  :)

**Comment by sasmaster** [comment]
Making PR + moved that header into includes.

**Comment by D7EAD** [comment]
> Good day!
> 
> I tried running the example https://github.com/D7EAD/liboai/blob/main/documentation/chat/examples/create_chat_completion.cpp
> 
> And I also encountered this bug on my configuration: Windows 10, MSYS2, g++ compiler, linked with the currently available curl lib from MSYS2. I was building liboai from the source code, and I received "signal SIGTRAP" at the end of the
> 
> ```
> liboai::netimpl::Session::~Session() noexcept { ... }
> ```
> 
> destructor. Thanks to sasmaster, looks like the source of the error was found ^_^ I tried the code from the corresponding pull request, and it worked without any errors.
> 
> However, the main reason I am mentioning this is because I met another consequent issue, with the help of updated code:
> 
> ```
> liboai::netimpl::Session::Perform(): Timeout was reached (E_CURLERROR:0x06).
> ```
> 
> It seems that 30 seconds isn't enough for OpenAI to respond in some cases.
> 
> To solve this issue, I increased the timeout numbers inside network.h:
> 
> ```
> Line 108: netimpl::components::Timeout{ 30000 },
> Line 116: netimpl::components::Timeout{ 30000 },
> ```
> 
> After doing this, the example worked perfectly fine, and I received a result!
> 
> It appears that at this exact minute there is high demand for the OpenAI servers, so I guess, it is not usual, that 30 seconds are not enough, which maybe explains the difficulties with the initial bug reproduction. Therefore, I suggest that these numbers be increased or at least made configurable from the client code somehow =)

Good idea. I'll push a change to `authorization.h` and `.cpp` to allow users to dynamically set their timeout values in seconds.
